packages/protocol/contracts/common/MetaTransactionWallet.sol

```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
   }
 
   /**
-   * @notice Executes a transaction on behalf of the signer.`
+  * relayer? TODO determine if we can predict this request (question for komenci folks)
+  * gatewayFee / gateWayFeeRecipient? (TODO: ask Contracts if better to omit these or add them for forward compatibility) 
+  */
+
+  /**
+   * @notice Executes a refundable meta-transaction on behalf of the signer.
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param maxGasPrice The maximum gas price the user is willing to pay.
+   * @param gasLimit The gas limit for entire relayed transaction.
+   * @param metaGasLimit The gas limit for just the meta-transaction.
+   * @param v The recovery id of the ECDSA signature of the meta-transaction.
+   * @param r Output value r of the ECDSA signature.
+   * @param s Output value s of the ECDSA signature.
+   * @return The return value of the meta-transaction execution.
+   */
+  function executeMetaTransactionWithRefund(
+    address destination,
+    uint256 value,
+    bytes calldata data,
+    uint256 maxGasPrice,
+    uint256 gasLimit,
+    uint256 metaGasLimit,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit == gasLeft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
+    require(address(this).balance >= gasLimit.mul(tx.gasprice).add(value), "insufficient balance");
+    require(msg.sender == tx.origin, "relayer must be EOA");
+    require(gasLimit <= block.gaslimit, "Impossible gas limit");
+    require(metaGasLimit < gasLimit, "metaGasLimit must be less than gasLimit");
+    if (data.length > 0) require(Address.isContract(destination), "Invalid contract address");
+
+    // checking gasCurrency is out of scope, would require pre-compile
+
+    address _signer = getMetaTransactionWithRefundSigner(
+      destination,
+      value,
+      data,
+      nonce,
+      maxGasPrice,
+      metaGasLimit,
+      v,
+      r,
+      s
+    );
+    require(_signer == signer, "Invalid meta-transaction signer");
+    nonce = nonce.add(1);
+
+    // TODO: ask Contracts
+    uint256 buffer1 = 4949; // TODO: determine this constant (gas required for operations after destination.call.value)
+    uint256 buffer2 = 4747; // TODO: determine this constant (gas required for operations after and including msg.sender.transfer)
+
+    if (address(this).balance >= metaGasLimit.add(buffer1).mul(tx.gasprice).add(value)) {
```

Comment:
hmmm okay so if there's not enough gas to execute the external call then we just won't do the call, but we will increment the nonce? 🤔 this seems kinda fishy to me but again i don't have perfect context

Reply:
```
// determine whether there is enough gas left 
// to execute the meta transaction
uint256 gasNeeded = metaGasLimit.add(buffer1);
bool enoughGas = gasLeft() >= gasNeeded;
// determine whether balance of wallet is enough
// to execute the meta transaction
uint256 currencyNeeded = gasNeeded.mul(tx.gasPrice).add(value);
bool enoughCurrency = address(this).balance >= currencyNeeded
// if enough gas & enough balance, execute transaction
if(enoughGas && enoughCurrency) {
```